db/upgrade.php

```diff
@@ -269,5 +225,54 @@ function xmldb_report_customsql_upgrade($oldversion) {
         upgrade_plugin_savepoint(true, 2021111600, 'report', 'customsql');
     }
 
-    return true;
+    if ($oldversion < 2022031801) {
+
+        // Define table report_customsql_categories to be created.
+        $table = new xmldb_table('report_customsql_categories');
+
+        // Adding fields to table report_customsql_categories.
+        $table->add_field('id', XMLDB_TYPE_INTEGER, '10', null, XMLDB_NOTNULL, XMLDB_SEQUENCE, null, null);
+        $table->add_field('name', XMLDB_TYPE_CHAR, '255', null, XMLDB_NOTNULL, null, null, null);
+
+        // If the 'name' field already exists, ensure its constraint is set to NOTNULL
+        if ($dbman->field_exists($table, $table->getField('name'))) {
+            $dbman->change_field_notnull($table, $table->getField('name'));
+        }
+
+        // Adding key to table report_customsql_categories.
+        $table->add_key('primary', XMLDB_KEY_PRIMARY, array('id'));
+
+        // Conditionally launch create table for report_customsql_categories.
+        if (!$dbman->table_exists($table)) {
+            $dbman->create_table($table);
+        }
+
+        // Define field categoryid to be added to report_customsql_queries.
+        $table = new xmldb_table('report_customsql_queries');
+        $field = new xmldb_field('categoryid', XMLDB_TYPE_INTEGER, '10', XMLDB_UNSIGNED, null, null, null, 'emailwhat');
+
+        // Conditionally launch add field categoryid.
+        if (! $dbman->field_exists($table, $field)) {
+            $dbman->add_field($table, $field);
+        }
+
+        // Add key (for the new field just added).
+        $key = new xmldb_key('categoryid', XMLDB_KEY_FOREIGN, array('categoryid'), 'report_customsql_categories', array('id'));
+        $dbman->add_key($table, $key);
+
+        // Create the default 'Miscellaneous' category.
+        $category = new stdClass();
+        $category->name = get_string('defaultcategory', 'report_customsql');
+        if (!$DB->record_exists('report_customsql_categories', array('name' => $category->name))) {
+            $category->id = $DB->insert_record('report_customsql_categories', $category);
+            // Update the existing query category ids, to move them into this category.
+            $sql = 'UPDATE {report_customsql_queries} SET categoryid =' . $category->id;
```

Comment:
Prior, this would be run unconditionally which would cause an error since [#L123](https://github.com/catalyst/moodle-report_customsql/blob/main-catalyst/db/upgrade.php#L123) would be skipped.

Reply:
I'm not sure I understand what you mean. Could you elaborate on this?

I can definitely see an issue if this was kept in the new upgrade block.

Reply:
My understanding is that sites that already have this plugin installed will have the `report_customsql_categories`   record already, so `if (!$DB->record_exists('report_customsql_categories', ...`   will be skipped, leaving `$category->id`   undefined.  This was causing [`$DB->execute($sql);`](https://github.com/catalyst/moodle-report_customsql/blob/main-catalyst/db/upgrade.php#L128) to throw an error.

Reply:
Correct. As pointed out by the other comments, I believe only the updates needed should be in the new block, instead of trying to recreate the table and the surrounding logic, etc.

Reply:
Agreed, my bad there. I've left the original block alone and created a new one with what we need to resolve the failed schema check